src/sql/src/kafka_util.rs

```diff
@@ -246,6 +250,132 @@ pub async fn test_config(
+/// timestamp is greater than or equal to the given timestamp for the
+/// partition. If no such message exists (or the Kafka broker is before
+/// 0.10.0), the current end offset is returned for the partition.
+///
+/// The provided `time_offset` option must be a non-zero number:
+/// * Non-Negative numbers will used as is (e.g. `1622659034343`)
+/// * Negative numbers will be translated to a timestamp in millis
+///   before now (e.g. `-10` means 10 millis ago)
+///
+/// If `time_offset` has not been configured, an empty Option is returned.
+pub async fn lookup_start_offsets(
+    broker: &str,
+    topic: &str,
+    client_options: &BTreeMap<String, String>,
+    with_options: &BTreeMap<String, Value>,
+) -> Result<Option<Vec<i64>>, anyhow::Error> {
+    let time_offset = with_options.get("time_offset");
+    if time_offset.is_none() {
+        return Ok(None);
+    } else if with_options.contains_key("start_offset") {
+        bail!("`start_offset` and `time_offset` cannot be set at the same time.")
+    }
+
+    // Validate and resolve `time_offset`.
+    let time_offset = match time_offset.unwrap() {
+        Value::Number(s) => match s.parse::<i64>() {
+            // Timestamp in millis *before* now (e.g. -10 means 10 millis ago)
+            Ok(ts) if ts < 0 => {
+                let now: i64 = SystemTime::now()
+                    .duration_since(SystemTime::UNIX_EPOCH)?
+                    .as_millis()
+                    .try_into()?;
+                let ts = now - ts.abs();
+                if ts <= 0 {
+                    bail!("Relative `time_offset` must be smaller than current system timestamp")
+                }
+                ts
+            }
+            // Timestamp in millis (e.g. 1622659034343)
+            Ok(ts) => ts,
+            _ => bail!("`time_offset` must be a number"),
+        },
+        _ => bail!("`time_offset` must be a number"),
+    };
+
+    // Lookup offsets
+    let mut config = rdkafka::ClientConfig::new();
+    config.set("bootstrap.servers", broker);
+    for (k, v) in client_options {
+        config.set(k, v);
+    }
+
+    let consumer: BaseConsumer<KafkaErrCheckContext> =
+        match config.create_with_context(KafkaErrCheckContext::default()) {
+            Ok(consumer) => consumer,
+            Err(e) => {
+                match e {
+                    rdkafka::error::KafkaError::ClientCreation(s) => {
+                        // Pass existing error back up.
+                        bail!(rdkafka::error::KafkaError::ClientCreation(s))
+                    }
+                    _ => bail!(e),
+                }
+            }
+        };
+
+    task::spawn_blocking({
+        let topic = topic.to_string();
+        move || {
+            // There cannot be more than i32 partitions
+            let num_partitions: i32 = consumer
+                .fetch_metadata(Some(&topic), Duration::from_secs(10))?
+                .topics()
+                .first()
```

Comment:
It looks like we'll want to check if `fetch_metadata` returns more than one result -- we do that in the [`sink_connector` code](https://github.com/MaterializeInc/materialize/blob/main/src/coord/src/sink_connector.rs#L72). I think you could also pull `get_partitions` into `kafka_util`, if you're interested!

Reply:
Thanks for the hint! Move to `kafka_util`.